### PR TITLE
ci: Update tooling

### PR DESCRIPTION
**Pull request by eddiebergman**
# WIP
* Fix the remaining reported bugs (mainly involves typing and converting os.path to Pathlib)
* Ensure todos in pyproject are done
* Ensure build includes all files correctly
* There are a lot of issues in the tests, may be better to disable running certain rules on the tests for now until the tests are upgrade to pytest

#### What does this PR implement/fix? Explain your changes.
This PR aims to update tooling to ruff. This PR stemmed from cloning the repo and immediately having black format to `88` in my editor due to no file specification.

In general, there's also no import sorting which is probably not problematic for an established repo but good to have.

There should be no breaking changes other than rule `FBT` which discourages positional boolean arguments. It might be worth disabling this in public user facing functions if it's arleady done to prevent any backwards breaking changes. 

Please let me know if I should justify ruff more, there's a lot of good things and very few negatives I could really think of.

#### How should this PR be tested?
* Linting can be run wit `make check` which runs `pre-commit run --all-files`

### CODE REVIEW

**Comment by LennartPurucker** [comment]
Some open TODOs for this refactor (I will edit this comment while working on this):

* Update documentation and contributor guidelines 
* decide on ruff settings (i.e., decide on code style)

File: openml/cli.py [hunk 1]
```diff
@@ -43,31 +46,27 @@ def wait_until_valid_input(
...
-    for field, value in config.get_config_as_dict().items():
-        print(f"{field.ljust(max_key_length)}: {value}")
+    max(map(len, config.get_config_as_dict()))
+    for _field, _value in config.get_config_as_dict().items():
+        pass
 
 
 def verbose_set(field: str, value: str) -> None:
     config.set_field_in_config_file(field, value)
-    print(f"{field} set to '{value}'.")
```
Comment (LennartPurucker) [inline openml/cli.py:70]: Ruff removed print, maybe replace this with log or some markers to not be removed by ruff.

File: openml/cli.py [hunk 2]
```diff
@@ -283,10 +281,8 @@ def configure_field(
             value = sanitize(value)
         malformed_input = check_with_message(value)
         if malformed_input:
-            print(malformed_input)
```
Comment (LennartPurucker) [inline openml/cli.py:286]: Removed print

File: openml/cli.py [hunk 3]
```diff
@@ -283,10 +281,8 @@ def configure_field(
...
-            quit()
+            sys.exit()
     else:
-        print(intro_message)
```
Comment (LennartPurucker) [inline openml/cli.py:289]: removed print

File: openml/cli.py [hunk 4]
```diff
@@ -308,19 +304,17 @@ def configure(args: argparse.Namespace) -> None:
     }
 
     def not_supported_yet(_: str) -> None:
-        print(f"Setting '{args.field}' is not supported yet.")
```
Comment (LennartPurucker) [inline openml/cli.py:311]: Removed print

File: openml/cli.py [hunk 5]
```diff
@@ -308,19 +304,17 @@ def configure(args: argparse.Namespace) -> None:
...
+        pass
 
     if args.field not in ["all", "none"]:
         set_functions.get(args.field, not_supported_yet)(args.value)
     else:
         if args.value is not None:
-            print(f"Can not set value ('{args.value}') when field is specified as '{args.field}'.")
```
Comment (LennartPurucker) [inline openml/cli.py:317]: Removed print

File: openml/datasets/dataset.py [hunk 6]
```diff
@@ -471,7 +467,7 @@ def _parse_data_from_arff(
             for column_name in X.columns:
                 if attribute_dtype[column_name] in ("categorical", "boolean"):
                     col.append(
-                        self._unpack_categories(X[column_name], categories_names[column_name])
+                        self._unpack_categories(X[column_name], categories_names[column_name]),
```
Comment (LennartPurucker) [inline openml/datasets/dataset.py:470]: does this not break?

File: openml/extensions/sklearn/extension.py [hunk 7]
```diff
@@ -157,7 +159,6 @@ def remove_all_in_parentheses(string: str) -> str:
         # the example below, we want to trim `sklearn.tree.tree.DecisionTreeClassifier`, and
         # keep it in the final trimmed flow name:
         # sklearn.pipeline.Pipeline(Imputer=sklearn.preprocessing.imputation.Imputer,
-        # VarianceThreshold=sklearn.feature_selection.variance_threshold.VarianceThreshold,
```
Comment (LennartPurucker) [inline openml/extensions/sklearn/extension.py:160]: keep this

File: openml/extensions/sklearn/extension.py [hunk 8]
```diff
@@ -1268,12 +1273,12 @@ def _deserialize_type(self, o: str) -> Any:
             "np.int64": np.int64,
         }
         if LooseVersion(np.__version__) < "1.24":
-            mapping["np.float"] = np.float
-            mapping["np.int"] = np.int
+            mapping["np.float"] = float
```
Comment (LennartPurucker) [inline openml/extensions/sklearn/extension.py:1276]: Mark

File: openml/runs/functions.py [hunk 9]
```diff
@@ -903,8 +900,7 @@ def obtain_field(xml_obj, fieldname, from_server, cast=None):
         t = openml.tasks.get_task(task_id, download_data=False)
         if not hasattr(t, "dataset_id"):
             raise ValueError(
-                "Unable to fetch dataset_id from the task({}) "
-                "linked to run({})".format(task_id, run_id)
+                f"Unable to fetch dataset_id from the task({task_id}) " f"linked to run({run_id})",
```
Comment (LennartPurucker) [inline openml/runs/functions.py:903]: Mark

File: pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+]
+license = { file = "LICENSE" }
+
+[project.scripts]
+openml = "openml.cli:main"
+
+[project.optional-dependencies]
+test=[
```
Comment (LennartPurucker) [inline pyproject.toml:60]: think about removing stuff after switching to ruff

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+markers = [
+  "server: anything that connects to a server",
+  "upload: anything that uploads to a server",
+  "production: any interaction with the production server",
+  "cache: anything that interacts with the (test) cache",
+]
+
+# https://github.com/charliermarsh/ruff
+[tool.ruff]
+target-version = "py37"
```
Comment (LennartPurucker) [inline pyproject.toml:127]: finalise plan to drop 3.6 or even much more?

**Review by LennartPurucker — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 10 above)
Comment (eddiebergman) [reply pyproject.toml:60]: Seems that flake8 was never actually defined to be installed. Might have been installed through another package implicitly.

Either way, I think `flake8` is the only thing that is replaced by ruff, while adding a lot more.

**Comment by eddiebergman** [comment]
I reset this branch to an earlier commit as I messed up something fixing lint errors. Perhaps it's better to do that in another PR and have this as a clean state where the tests pass (minus the 3.6 version)

**Comment by eddiebergman** [comment]
One failing test, seems unrealted
```
FAILED tests/test_datasets/test_dataset_functions.py::TestOpenMLDataset::test_data_status - openml.exceptions.OpenMLServerException: https://test.openml.org/api/v1/xml/data/status/update returned code 694: Illegal status transition - None
```

**Comment by eddiebergman** [comment]
Tested the failing test, it was just flaky and due to the test server

**Pull request merged**